src/turnkeyml/analyze/discover.py

```diff
@@ -0,0 +1,186 @@
+
+    def fire(
+        self,
+        state: fs.State,
+        input: str = "",
+        targets: Optional[List[str]] = None,
+        script_args: str = "",
+        max_depth: int = default_max_depth,
+    ):
+        if not input.endswith(".py"):
+            raise exp.ArgError(
+                "Inputs to the `discover` stage must by python scripts "
+                f"(.py files), got {input}",
+            )
+
+        if targets is None:
+            targets_to_use = []
+        else:
+            targets_to_use = targets
+
+        tracer_args = TracerArgs(
+            input=input,
+            script_args=script_args,
+            targets=targets_to_use,
+            max_depth=max_depth,
+        )
+
+        # Discover the models in the python script by executing it with
+        # a tracer enabled
+        models_found = evaluate_script(tracer_args)
+
+        # Count the amount of build-able model invocations discovered
+        # If there is only 1, pass it to the next build stage. Otherwise,
+        # print all the invocations and suggest that the user select one.
+        count = 0
+        for model_info in models_found.values():
+            for (
+                invocation_hash,
+                invocation_info,
+            ) in model_info.unique_invocations.items():
+                count += 1
+
+                # Set the same status for all invocations at first
+                # The next code block will be responsible for the selected
+                # invocation.
+
+                invocation_info.status_message = (
+                    "Discovered; select with `-i "
+                    f"{os.path.basename(model_info.file)}::{invocation_hash}"
+                )
+                invocation_info.status_message_color = printing.Colors.OKCYAN
+
+        for model_info in models_found.values():
+            for invocation_info in model_info.unique_invocations.values():
+                if invocation_info.is_target or (
+                    len(targets_to_use) == 0 and count > 1
+                ):
+                    if len(targets_to_use) == 0 and count > 1:
+                        invocation_info.auto_selected = True
+
+                    # Save stats about the model
+                    state.save_stat(
+                        fs.Keys.HASH,
+                        model_info.hash,
+                    )
+                    state.save_stat(
+                        fs.Keys.MODEL_NAME,
+                        tracer_args.script_name,
+                    )
+                    state.save_stat(
+                        fs.Keys.PARAMETERS,
+                        model_info.params,
+                    )
+
+                    state.save_stat(
+                        fs.Keys.CLASS,
+                        type(model_info.model).__name__,
+                    )
+
+                    state.results = model_info.model
+                    # FIXME: we should be able to get rid of state.model now
```

Comment:
Recommend to either fix or add an issue here.

Reply:
Documented in #168